### ISSUE TITLE
ActivationException: This operation is only valid on generic types.

### ISSUE THREAD

**Issue by QuantumHive** [simpleinjector/SimpleInjector#471]
Related reference see this [SO post](https://stackoverflow.com/questions/46777396/pagination-as-cross-cutting-concern-in-cqrs-with-simpleinjector).
To reproduce, see [this gist](https://gist.github.com/QuantumHive/697c6a7cdb680f66d9e52f5740154e21).

When using `RegisterConditional` on an open generic interface like

    IQueryHandler<TQuery, TResult>

which if `true` returns a dynamically created closed type of an open generic class like

    PaginationQueryHandlerInterceptor<TQuery, TModel>

the following unclear exception is thrown:

```
System.InvalidOperationException occurred
  HResult=0x80131509
  Message=The configuration is invalid. Creating the instance for type Program failed. Value cannot be null.
Parameter name: type
  Source=<Cannot evaluate the exception source>
  StackTrace:
   at SimpleInjector.InstanceProducer.VerifyExpressionBuilding()
   at SimpleInjector.Container.VerifyThatAllExpressionsCanBeBuilt(InstanceProducer[] producersToVerify)
   at SimpleInjector.Container.VerifyThatAllExpressionsCanBeBuilt()
   at SimpleInjector.Container.VerifyInternal(Boolean suppressLifestyleMismatchVerification)
   at SimpleInjector.Container.Verify()

Inner Exception 1:
ActivationException: Value cannot be null.
Parameter name: type

Inner Exception 2:
ArgumentNullException: Value cannot be null.
```

I think this exception message (and the one from the SO post) should be more clear?

**Comment by dotnetjunkie** [comment]
Can you post the stack traces of the inner exceptions as well?

**Comment by dotnetjunkie** [comment]
> I think this exception message (and the one from the SO post) should be more clear

What do you suggest the exception message should be?

**Comment by QuantumHive** [comment]
> Can you post the stack traces of the inner exceptions as well?

```
   at System.Reflection.IntrospectionExtensions.GetTypeInfo(Type type)
   at SimpleInjector.Internals.GenericTypeBuilder..ctor(Type closedGenericBaseType, Type openGenericImplementation)
   at SimpleInjector.Internals.GenericTypeBuilder.MakeClosedImplementation(Type closedAbstraction, Type openImplementation)
   at SimpleInjector.Internals.GenericRegistrationEntry.OpenGenericToInstanceProducerProvider.MatchesServiceType(Type serviceType)
   at System.Linq.Utilities.<>c__DisplayClass1_0`1.<CombinePredicates>b__0(TSource x)
   at System.Linq.Enumerable.WhereListIterator`1.GetCount(Boolean onlyIfCheap)
   at System.Linq.Enumerable.Count[TSource](IEnumerable`1 source)
   at SimpleInjector.Container.GetNumberOfConditionalRegistrationsFor(Type serviceType)
   at SimpleInjector.Container.ThrowParameterTypeMustBeRegistered(InjectionTargetInfo target)
   at SimpleInjector.Advanced.DefaultDependencyInjectionBehavior.GetInstanceProducer(InjectionConsumerInfo consumer, Boolean throwOnFailure)
   at SimpleInjector.ContainerOptions.GetInstanceProducerFor(InjectionConsumerInfo consumer)
   at SimpleInjector.Registration.BuildConstructorParameters(ConstructorInfo constructor)
   at SimpleInjector.Registration.BuildNewExpression()
   at SimpleInjector.Registration.BuildTransientExpression()
   at SimpleInjector.InstanceProducer.BuildExpressionInternal()
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at SimpleInjector.InstanceProducer.BuildExpression()
```

> What do you suggest the exception message should be?

The message says the configuration is invalid. So as a consumer I'm trying to figure out why the configuration is invalid. The message tells us that *Value cannot be null. Parameter name: type*. But it's not obvious which parameter has the value `null`. So as a consumer I'm at a loss trying to fix the configuration without actually diving in the source code of Simple Injector. I'm cannot suggest anything concrete, because my knowledge ceases at this point. I'm posting this so that perhaps the framework's exception messages could be improved, or maybe this is a bug?

**Comment by dotnetjunkie** [comment]
You are right, there is an unfortunate bug in the error reporting part of the system. This causes the real problem (there is no matched registration) to be hidden. This happens solely in the case that there is a registration for an conditional open-generic type using a type factory method.

The shortest code to reproduce the issue is the following:

``` c#
interface IGeneric<T> { }
class Consumer { public Consumer(IGeneric<T> g) { } }

var container = new Container();
container.RegisterConditional(typeof(IGeneric<>), c => null, Lifestyle.Singleton, c => false);

container.GetInstance<Consumer>();
```

**Comment by QuantumHive** [comment]
@dotnetjunkie relating to the SO post and [this answer](https://stackoverflow.com/a/46786765/2040663), I would suggest the following:

Perhaps Simple Injector could be improved by telling the developer that the exception originates from the `RegisterConditional` predicate and that he/she should check it's implementation for any errors?

The same applies to my gist example, because the *value cannot be null* exception is caused by a bad implementation of the predicate:

```
bool Predicate(PredicateContext context) =>
	context.ServiceType.GetGenericArguments()[0].GetInterfaces()
		.Any(i => i.GetGenericTypeDefinition() == typeof(PaginatedQuery<>));
```

**Comment by dotnetjunkie** [comment]
> Perhaps Simple Injector could be improved by telling the developer that the exception originates from the RegisterConditional predicate 

I would say obviously. Typically, you will get an error similar to the following:

> 1 conditional registration for IGeneric<T> exists that is applicable to IGeneric<Int32>, but its supplied predicate didn't return true when provided with the contextual information for Consumer<IGeneric<Int32>>

When the bug is fixed, you can expect similar information.

Currently, the thing that goes wrong is the calculation of the number of conditional registrations that apply. Notice the error message starts with a number. This is calculated, and is only calculated in case of an error. But this calculation fails, because it didn't consider type factories :'(.

**Comment by QuantumHive** [comment]
> there is an unfortunate bug in the error reporting part of the system. This causes the real problem (there is no matched registration) to be hidden.

I guess [two flies in one hit](http://nl.bab.la/woordenboek/nederlands-engels/twee-vliegen-in-%C3%A9%C3%A9n-klap-slaan)?

**Comment by dotnetjunkie** [comment]
Nah, just one fly, but still pretty fly for a wifly :)

**Comment by dotnetjunkie** [comment]
Fixed in [v4.0.12](https://www.nuget.org/packages/SimpleInjector/4.0.12).
